Geometry/MTDNumberingBuilder/plugins/CmsMTDConstruction.cc

```diff
@@ -1,42 +1,31 @@
 #include "Geometry/MTDNumberingBuilder/plugins/CmsMTDConstruction.h"
 #include "Geometry/MTDNumberingBuilder/plugins/ExtractStringFromDDD.h"
-#include "DataFormats/DetId/interface/DetId.h"
 
 #include "DataFormats/ForwardDetId/interface/BTLDetId.h"
 #include "DataFormats/ForwardDetId/interface/ETLDetId.h"
 
-void CmsMTDConstruction::buildComponent(DDFilteredView& fv, GeometricTimingDet* mother, std::string attribute) {
-  //
-  // at this level I check whether it is a merged detector or not
-  //
-
-  GeometricTimingDet* det = new GeometricTimingDet(&fv, theCmsMTDStringToEnum.type(fv.name()));
-
-  const std::string part_name = fv.name().substr(0, 7);
-
-  if (theCmsMTDStringToEnum.type(part_name) == GeometricTimingDet::BTLModule) {
-    bool dodets = fv.firstChild();
-    while (dodets) {
-      buildBTLModule(fv, det, attribute);
-      dodets = fv.nextSibling();
-    }
-    fv.parent();
-  } else if (theCmsMTDStringToEnum.type(part_name) == GeometricTimingDet::ETLModule) {
-    bool dodets = fv.firstChild();
-    while (dodets) {
-      buildETLModule(fv, det, attribute);
-      dodets = fv.nextSibling();
-    }
-    fv.parent();
-  } else {
-    throw cms::Exception("MTDConstruction") << "woops got: " << part_name << std::endl;
-  }
+#include "DataFormats/Math/interface/deltaPhi.h"
 
-  mother->addComponent(det);
+bool CmsMTDConstruction::mtdOrderZ(const GeometricTimingDet* a, const GeometricTimingDet* b) {
+  bool order = (a->translation().z() == b->translation().z()) ? a->translation().rho() < b->translation().rho()
+                                                              : a->translation().z() < b->translation().z();
+  return order;
+}
+
+bool CmsMTDConstruction::mtdOrderRR(const GeometricTimingDet* a, const GeometricTimingDet* b) {
+  MTDDetId id1(a->geographicalId());
+  MTDDetId id2(b->geographicalId());
+  return id1.mtdRR() < id2.mtdRR();
+}
+
+bool CmsMTDConstruction::mtdOrderPhi(const GeometricTimingDet* a, const GeometricTimingDet* b) {
+  MTDDetId id1(a->geographicalId());
+  MTDDetId id2(b->geographicalId());
+  return (id1.mtdRR() == id2.mtdRR()) && (angle0to2pi::make0To2pi(a->phi()) < angle0to2pi::make0To2pi(b->phi()));
 }
 
 void CmsMTDConstruction::buildBTLModule(DDFilteredView& fv, GeometricTimingDet* mother, const std::string& attribute) {
-  GeometricTimingDet* det = new GeometricTimingDet(&fv, theCmsMTDStringToEnum.type(fv.name()));
+  GeometricTimingDet* det = new GeometricTimingDet(&fv, theCmsMTDStringToEnum.type(fv.name().substr(0, 7)));
```

Comment:
magic number 7 should be named constant and reused below

Reply:
not the only one, but annoying enough to be dealt with, I agree